src/System.Memory/src/System/Buffers/ReadOnlyBuffer.cs

```diff
@@ -0,0 +1,377 @@
+        public ReadOnlyBuffer<T> Slice(long offset)
+        {
+            if (offset == 0) return this;
+
+            var begin = Seek(BufferStart, BufferEnd, offset, false);
+            return SliceImpl(begin, BufferEnd);
+        }
+
+        public override string ToString()
+        {
+            if (typeof(T) == typeof(byte))
+            {
+                if (this is ReadOnlyBuffer<byte> bytes)
+                {
+                    var sb = new StringBuilder();
+                    foreach (var buffer in bytes)
+                    {
+                        SpanLiteralExtensions.AppendAsLiteral(buffer.Span, sb);
+                    }
+                    return sb.ToString();
+                }
+            }
+            return base.ToString();
+        }
+
+        /// <summary>
+        /// Returns an enumerator over the <see cref="ReadOnlyBuffer{T}"/>
+        /// </summary>
+        public Enumerator GetEnumerator()
+        {
+            return new Enumerator(this);
+        }
+
+        public SequencePosition GetPosition(SequencePosition origin, long offset)
+        {
+            if (offset < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(offset));
+            }
+            return Seek(origin, BufferEnd, offset, false);
+        }
+
+        public bool TryGet(ref SequencePosition position, out ReadOnlyMemory<T> data, bool advance = true)
+        {
+            var result = TryGetBuffer(position, End, out data, out var next);
+            if (advance)
+            {
+                position = next;
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// An enumerator over the <see cref="ReadOnlyBuffer{T}"/>
+        /// </summary>
+        public struct Enumerator
+        {
+            private readonly ReadOnlyBuffer<T> _readOnlyBuffer;
+            private SequencePosition _next;
+            private ReadOnlyMemory<T> _currentMemory;
+
+            /// <summary>
+            ///
+            /// </summary>
+            public Enumerator(ReadOnlyBuffer<T> readOnlyBuffer)
+            {
+                _readOnlyBuffer = readOnlyBuffer;
+                _currentMemory = default;
+                _next = readOnlyBuffer.Start;
+            }
+
+            /// <summary>
+            /// The current <see cref="Memory{Byte}"/>
+            /// </summary>
+            public ReadOnlyMemory<T> Current => _currentMemory;
+
+            /// <summary>
+            /// Moves to the next <see cref="Memory{Byte}"/> in the <see cref="ReadOnlyBuffer{T}"/>
+            /// </summary>
+            /// <returns></returns>
+            public bool MoveNext()
+            {
+                if (_next.Segment == null)
+                {
+                    return false;
+                }
+
+                return _readOnlyBuffer.TryGet(ref _next, out _currentMemory);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
@jkotas, @AndyAyersMS is the use of Aggressiveness necessary here?

Reply:
My recommendation would to not use any inlining attributes at all initially.

You can selectively add them when you have performance scenarios where you can better assess cost and benefit.

Reply:
They were added as part of kestrel performance work when these types were in corefxlab.

Reply:
Number of these look pretty questionable to me though. You can always do microbenchmark "performance work" by blindly adding AggresiveInlining close to everywhere - but it does hurt startup and often also hurts throughput for real word loads because of the code does not fit in the caches.

Reply:
At the time of applying attributes improved kestrel plaintext results.

Reply:
But methods did increase in size since that time so maybe re-investigating or extracting main code path is the right thing to do here.

Reply:
For inlining, it's about examining balance, not just looking at benefit -- most inlines will improve performance in some scenarios, but at what cost?

Bigger methods take longer to jit and create more code. So it is good to develop a more complete picture.

Reply:
We should remove this *after* we're consuming the code from corefx. We've already regressed the performance by > 10% because of small tweaks so I'm hesitant to make changes that may affect the performance when we're just transferring code.

Reply:
Open issue so that this is not forgotten.

Reply:
https://github.com/dotnet/corefx/issues/26740